core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/SplitsMapping.java

```diff
@@ -0,0 +1,299 @@
+                        // no source splits for give sourcePartitionId
+                        targetSplitsMapBuilder.put(sourcePartitionId, innerEntry.getValue().build());
+                    }
+                }
+
+                for (Map.Entry<Integer, List<Split>> innerEntry : originalSplitMap.entrySet()) {
+                    Integer sourcePartitionKey = innerEntry.getKey();
+                    if (updatesSplitsMap.containsKey(sourcePartitionKey)) {
+                        // already handled above
+                        continue;
+                    }
+                    // reuse source immutable split list
+                    targetSplitsMapBuilder.put(sourcePartitionKey, innerEntry.getValue());
+                }
+                immutableSplitsBuilder.put(planNodeId, targetSplitsMapBuilder.build());
+            }
+
+            for (Map.Entry<PlanNodeId, Map<Integer, ImmutableList.Builder<Split>>> entry : updates.entrySet()) {
+                PlanNodeId planNodeId = entry.getKey();
+                if (originalMapping.splits.containsKey(planNodeId)) {
+                    // handled above
+                    continue;
+                }
+
+                ImmutableMap.Builder<Integer, List<Split>> targetSplitsMapBuilder = ImmutableMap.builder();
+                for (Map.Entry<Integer, ImmutableList.Builder<Split>> innerEntry : entry.getValue().entrySet()) {
+                    targetSplitsMapBuilder.put(innerEntry.getKey(), innerEntry.getValue().build());
+                }
+                immutableSplitsBuilder.put(planNodeId, targetSplitsMapBuilder.build());
+            }
+
+            return new SplitsMapping(immutableSplitsBuilder.buildOrThrow());
+        }
+    }
+
+    private static class NewBuilder
+            implements Builder
+    {
+        private final Map<PlanNodeId, Map<Integer, ImmutableList.Builder<Split>>> splitsBuilder = new HashMap<>();
+
+        @Override
+        public Builder addSplits(PlanNodeId planNodeId, int partitionId, List<Split> splits)
+        {
+            splitsBuilder.computeIfAbsent(planNodeId, ignored -> new HashMap<>())
+                    .computeIfAbsent(partitionId, ignored -> ImmutableList.builder())
+                    .addAll(splits);
+            return this;
+        }
+
+        @Override
+        public SplitsMapping build()
+        {
+            ImmutableMap.Builder<PlanNodeId, Map<Integer, List<Split>>> immutableSplitsBuilder = ImmutableMap.builder();
+            for (Map.Entry<PlanNodeId, Map<Integer, ImmutableList.Builder<Split>>> entry : splitsBuilder.entrySet()) {
+                immutableSplitsBuilder.put(entry.getKey(),
+                        entry.getValue().entrySet().stream().collect(toImmutableMap(
+                                Map.Entry::getKey,
+                                innerEntry -> innerEntry.getValue().build())));
+            }
+            return new SplitsMapping(immutableSplitsBuilder.buildOrThrow());
```

Comment:
nit: this yields to streams nicely. the following woul be equally hard to follow

```
return new SplitsMapping(splitsBuilder.entrySet().stream()
                    .collect(toImmutableMap(
                            Map.Entry::getKey,
                            planNodeMapping -> planNodeMapping.getValue().entrySet().stream()
                                    .collect(toImmutableMap(
                                            Map.Entry::getKey,
                                            sourcePartitionMapping -> sourcePartitionMapping.getValue().build())))));
```

Reply:
indeed